haystack/document_stores/memory.py

```diff
@@ -146,6 +172,25 @@ def write_documents(
                     )
                     continue
             self.indexes[index][document.id] = document
+            modified_documents += 1
+
+        if self.use_bm25 is True and modified_documents > 0:
+            self.update_bm25(index=index)
+
+    def update_bm25(self, index: Optional[str] = None):
+        """
+        Updates the BM25 sparse representation in the the document store.
+
+        :param index: Index name for which the BM25 representation is to be updated. If set to None, the default self.index is used.
+        """
+        index = index or self.index
+        tokenizer = re.compile(self.bm25_tokenization_regex).findall
```

Comment:
[I'm sure this is a bit of an early optimization, but here it goes :smile:]

I've seen you are using this regex in at least two places. We could speed up this a tiny bit by saving the compiled regex. At first I though we should just store the compile regex in the init method, but then, if someone tries to set it directly with `docstore.bm25_tokenization_regex = "a_new_regex"`, stuff will break.

So how about using a property?

```python
@property
def bm25_tokenization_regex(self):
    return self._tokenizer

@bm25_tokenization_regex.setter
def bm25_tokenization_regex(self, regex_string: str):
    self._tokenizer = re.compile(regex_string).findall
```
(untested)

 By adding these two method somewhere in the class, you can simply remove this line and use `self.bm25_tokenization_regex` directly instead of `tokenizer` in the line below.

Reply:
I like the idea!